### PR TITLE
Add specialized Ethereum receipt bloom filter

### PR DESCRIPTION
**Pull request by sheapwilliams**
- Reference Section 4.3.1 of the Yellow paper for details
- Enables searching receipt trie for topics and addresses

This is sitting on top of #31 , so is currently marked draft

### CODE REVIEW

File: include/monad/core/receipt.hpp [hunk 1]
```diff
@@ -17,15 +17,21 @@ struct Receipt
...
+        friend bool operator==(const Log &, const Log &) = default;
     };
 
     Bloom bloom{};
     uint64_t status{};
     uint64_t gas_used{};
     Transaction::Type type{};
     std::vector<Log> logs;
+
+    void add_log(Receipt::Log const &l);
```
Comment (zander-xyz) [inline include/monad/core/receipt.hpp:30]: any concern with having `logs` be a public member?

Comment (sheapwilliams) [reply include/monad/core/receipt.hpp:30]: None from me.  They are pretty simple structures.  No complicated invariants.  

Bloom is the one that has an invariant on the log list.  We can hide it and then unhide it again by providing getters/setters...but that seems overwrought to me.

Open to other thoughts here tho.

Comment (zander-xyz) [reply include/monad/core/receipt.hpp:30]: Nope, seems reasonable to me.

File: include/monad/core/receipt.hpp [hunk 2]
```diff
@@ -17,15 +17,21 @@ struct Receipt
         byte_string data{};
         std::vector<bytes32_t> topics{};
         address_t address{};
+
+        friend bool operator==(const Log &, const Log &) = default;
```
Comment (jhunsaker) [inline include/monad/core/receipt.hpp:21]: isn't there a default anyways?

File: src/monad/core/receipt.cpp [hunk 3]
```diff
@@ -0,0 +1,33 @@
+#include <ethash/keccak.hpp>
+#include <monad/core/receipt.hpp>
```
Comment (jhunsaker) [inline src/monad/core/receipt.cpp:2]:
```
#include <monad/core/receipt.hpp>

#include <ethash/keccak.hpp>
```

(don't group unrelated headers)

File: src/monad/core/receipt.cpp [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+        const uint16_t bit =
+            be16toh(reinterpret_cast<uint16_t const *>(h.bytes)[i]) & 2047;
+        const auto byte = 255 - bit / 8;
+        bloom[byte] |= 1 << (bit & 7);
+    }
+}
+
+void populate_bloom(Receipt::Bloom &b, Receipt::Log const &l)
+{
+    set_3_bits(b, byte_string_view{l.address.bytes, sizeof(l.address.bytes)});
```
Comment (jhunsaker) [inline src/monad/core/receipt.cpp:21]: `to_byte_string_view(l.address.bytes)`

File: src/monad/core/receipt.cpp [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    for (auto const &i : l.topics) {
+        set_3_bits(b, byte_string_view{i.bytes, sizeof(bytes32_t)});
```
Comment (jhunsaker) [inline src/monad/core/receipt.cpp:23]: same here

File: src/monad/core/receipt.cpp [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+#include <endian.h>
+
+MONAD_NAMESPACE_BEGIN
+
+void set_3_bits(Receipt::Bloom &bloom, byte_string_view const bytes)
+{
+    auto const h{ethash::keccak256(bytes.cbegin(), bytes.size())};
+    for (auto i = 0; i < 3; ++i) {
+        const uint16_t bit =
+            be16toh(reinterpret_cast<uint16_t const *>(h.bytes)[i]) & 2047;
```
Comment (jhunsaker) [inline src/monad/core/receipt.cpp:13]: we use the `intx` function `to_big_endian` in a lot of places - should we use it here consistency? I'm not sure how `bswap` compares against `be16toh` in this case (for our little endian builds). If `be16toh` is "faster" than `bswap` we should note that somewhere for future guidance.

File: include/monad/core/receipt.hpp (hunk 2 above)
Comment (sheapwilliams) [reply include/monad/core/receipt.hpp:21]: I had to add this after gtest functions complained.  Apparently not.

File: src/monad/core/receipt.cpp (hunk 4 above)
Comment (sheapwilliams) [reply src/monad/core/receipt.cpp:21]: Sure.

File: src/monad/core/receipt.cpp (hunk 6 above)
Comment (sheapwilliams) [reply src/monad/core/receipt.cpp:13]: This is the opposite - the bloom filter as an array of bytes is big endian.  We need a number out of it, so are converting to host byte order to do those calculations.  I think this uses the same `bswap` instructions under the hood, but I'll check the assembly.

Comment (sheapwilliams) [reply src/monad/core/receipt.cpp:13]: Discussed off-line.  `to_big_endian` implements the correct thing here, but the `to_` part carries semantic information that would be plain wrong.  

Changed in #37 and added comment to point out that dissonance.

File: src/monad/core/receipt.cpp (hunk 4 above)
Comment (sheapwilliams) [reply src/monad/core/receipt.cpp:21]: #37

**Pull request merged**